smart_contracts/contracts/explorer/faucet/src/lib.rs

```diff
@@ -1,39 +1,493 @@
+        FaucetError::InvalidFaucetPurse,
+    );
+
+    system::transfer_from_purse_to_account(faucet_purse, target, amount, id)
+        .unwrap_or_revert_with(FaucetError::FailedToTransfer);
+}
+
+fn get_distribution_amount_debounced() -> U512 {
+    let available_amount_uref = get_uref_with_user_errors(
+        AVAILABLE_AMOUNT,
+        FaucetError::MissingAvailableAmount,
+        FaucetError::InvalidAvailableAmount,
+    );
+
+    let available_amount: U512 = read_with_user_errors(
+        available_amount_uref,
+        FaucetError::MissingAvailableAmount,
+        FaucetError::InvalidAvailableAmount,
+    );
+
+    let remaining_amount_uref = get_uref_with_user_errors(
+        REMAINING_AMOUNT,
+        FaucetError::MissingRemainingAmount,
+        FaucetError::InvalidRemainingAmount,
+    );
+
+    let remaining_amount: U512 = read_with_user_errors(
+        remaining_amount_uref,
+        FaucetError::MissingRemainingAmount,
+        FaucetError::InvalidRemainingAmount,
+    );
+
+    let distributions_per_interval_uref = get_uref_with_user_errors(
+        DISTRIBUTIONS_PER_INTERVAL,
+        FaucetError::MissingDistributionsPerInterval,
+        FaucetError::InvalidDistributionsPerInterval,
+    );
+
+    let distributions_per_interval: u64 = read_with_user_errors(
+        distributions_per_interval_uref,
+        FaucetError::MissingDistributionsPerInterval,
+        FaucetError::InvalidDistributionsPerInterval,
+    );
 
-    let amount: U512 = runtime::get_named_arg(ARG_AMOUNT);
+    let distribution_amount =
+        Ratio::new(available_amount, U512::from(distributions_per_interval)).to_integer();
 
-    // Maybe we will decide to allow multiple funds up until some maximum value.
-    let already_funded = runtime::get_key(&account_hash.to_formatted_string()).is_some();
+    match remaining_amount.cmp(&distribution_amount) {
+        Ordering::Equal | Ordering::Greater => distribution_amount,
+        Ordering::Less => remaining_amount,
+    }
+}
+
+fn get_distribution_amount() -> U512 {
+    let available_amount_uref = get_uref_with_user_errors(
+        AVAILABLE_AMOUNT,
+        FaucetError::MissingAvailableAmount,
+        FaucetError::InvalidAvailableAmount,
+    );
+
+    let available_amount: U512 = read_with_user_errors(
+        available_amount_uref,
+        FaucetError::MissingAvailableAmount,
+        FaucetError::InvalidAvailableAmount,
+    );
+
+    let distributions_per_interval_uref = get_uref_with_user_errors(
+        DISTRIBUTIONS_PER_INTERVAL,
+        FaucetError::MissingDistributionsPerInterval,
+        FaucetError::InvalidDistributionsPerInterval,
+    );
+
+    let distributions_per_interval: u64 = read_with_user_errors(
+        distributions_per_interval_uref,
+        FaucetError::MissingDistributionsPerInterval,
+        FaucetError::InvalidDistributionsPerInterval,
+    );
+
+    Ratio::new(available_amount, U512::from(distributions_per_interval)).to_integer()
```

Comment:
```suggestion
    available_amount / U512::from(distributions_per_interval)
```
then we can remove the `num` crate dependency.